Dependencies/xr/Source/ARCore/XR.cpp

```diff
@@ -642,6 +655,158 @@ namespace xr
         return m_impl->TryInitialize();
     }
 
+    arcana::background_dispatcher<64> scheduler;
+    arcana::task<bool, std::exception_ptr> System::IsSessionSupportedAsync(SessionType sessionType)
+    {
+        // Currently only AR is supported on Android
+        if (sessionType == SessionType::IMMERSIVE_AR)
+        {
+            // Spin up a background thread to own the polling check.
+            arcana::task_completion_source<bool, std::exception_ptr> tcs;
+            arcana::make_task(scheduler, arcana::cancellation::none(), [sessionType, tcs]() mutable
+            {
+                // Query ARCore to check if AR sessions are supported.
+                // If not yet installed then poll supported status up to 100 times over 20 seconds.
+                for (int i = 0; i < 100; i++)
+                {
+                    ArAvailability arAvailability{};
+                    ArCoreApk_checkAvailability(GetEnvForCurrentThread(), GetAppContext(), &arAvailability);
+                    switch (arAvailability)
+                    {
+                        case AR_AVAILABILITY_SUPPORTED_APK_TOO_OLD:
+                        case AR_AVAILABILITY_SUPPORTED_INSTALLED:
+                        case AR_AVAILABILITY_SUPPORTED_NOT_INSTALLED:
+                            tcs.complete(true);
+                            break;
+                        case AR_AVAILABILITY_UNKNOWN_CHECKING:
+                            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+                            break;
+                        default:
+                            tcs.complete(false);
+                            break;
+                    }
+
+                    if (tcs.completed())
+                    {
+                        break;
+                    }
+                }
+
+                if (!tcs.completed())
+                {
+                    tcs.complete(false);
+                }
+            });
+
+            return tcs.as_task();
+        }
+
+        // VR and inline sessions are not supported at this time.
+        return arcana::task_from_result<std::exception_ptr>(false);
+    }
+
+    bool CheckARCoreInstallStatus(bool requestInstall)
```

Comment:
This is a global function. I would put this in an anonymous namespace like the other helper functions (e.g. `LoadShader`) so that it is not visible to anything outside this file, and can't conflict with other globals.

Reply:
Done